### PR TITLE
Implement @std/seq's recurse functionality in the AVM

### PR DESCRIPTION
**Pull request by dfellis**
Resolves \#265

### CODE REVIEW

File: avm/src/vm/memory.rs [hunk 1]
```diff
@@ -5,7 +5,7 @@ use crate::vm::program::Program;
 
 // -2^63
 pub const CLOSURE_ARG_MEM_START: i64 = -9223372036854775808;
-pub const CLOSURE_ARG_MEM_END: i64 = CLOSURE_ARG_MEM_START + 3;
+pub const CLOSURE_ARG_MEM_END: i64 = CLOSURE_ARG_MEM_START + 9001; // TODO: IT'S OVER 9000!
```
Comment (depombo) [inline avm/src/vm/memory.rs:8]: can we add some more context about why this is 9001 right now?

File: avm/src/vm/opcode.rs [hunk 2]
```diff
@@ -2468,7 +2468,37 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
...
+      let mut hand_mem = mem.write().await;
+      let mut hm = hand_mem.fork();
+      let (a, b) = hm.addr_to_idxs(args[0]);
+      hm.set_addr(CLOSURE_ARG_MEM_START + 1, a, b);
+      let (a, b) = hm.addr_to_idxs(args[1]);
+      hm.set_addr(CLOSURE_ARG_MEM_START + 2, a, b);
+      let slf = hm.read_fractal(args[0]);
+      let (mut seq, _) = hm.read_either_idxs(slf[0].0, slf[0].1 as usize);
+      let recurse_fn = HandlerFragment::new(slf[1].1, 0);
+      if seq[0].1 < seq[1].1 {
```
Comment (depombo) [inline avm/src/vm/opcode.rs:2482]: this is not really part of this PR, but I'm not a big fan of tuples just because they hurt legibility. also what does `slf` stand for?

File: compiler/src/ammtoaga.ts [hunk 3]
```diff
@@ -135,9 +135,28 @@ const getHandlersMem = (handlers: LPNode[]) => handlers
     return handlerMem
   })
 
-const closuresFromDeclaration = (declaration: LPNode, closureMem: object, eventDecs: object) => {
+const closuresFromDeclaration = (
+  declaration: LPNode,
+  closureMem: object,
+  eventDecs: object,
+  addressMap: object,
+  argRerefOffset: number,
```
Comment (depombo) [inline compiler/src/ammtoaga.ts:143]: can we leave a short comment about how argRerefOffset works?

**Review by depombo — APPROVED** [review]
Looks great, just minor comments!

File: avm/src/vm/memory.rs (hunk 1 above)
Comment (dfellis) [reply avm/src/vm/memory.rs:8]: Because it's over 9000!?

What!? 9000!?

File: compiler/src/ammtoaga.ts (hunk 3 above)
Comment (dfellis) [reply compiler/src/ammtoaga.ts:143]: Sure

File: avm/src/vm/opcode.rs (hunk 2 above)
Comment (dfellis) [reply avm/src/vm/opcode.rs:2482]: So, the tuples are the core of the memory model now. Rather than add some more memory model function calls, I just dug into it in just this one opcode as a trade-off.

And `slf` is `self` but the Rust compiler wouldn't let me call it that because the `self` variable is a reserved keyword only allowed inside of `impl` blocks for some insane reason.

File: avm/src/vm/memory.rs (hunk 1 above)
Comment (dfellis) [reply avm/src/vm/memory.rs:8]: I can add more, I just couldn't fit it after the constant so it'll have to be a new line.

File: avm/src/vm/opcode.rs (hunk 2 above)
Comment (depombo) [reply avm/src/vm/opcode.rs:2482]: > So, the tuples are the core of the memory model now. Rather than add some more memory model function calls, I just dug into it in just this one opcode as a trade-off.
> 
> And `slf` is `self` but the Rust compiler wouldn't let me call it that because the `self` variable is a reserved keyword only allowed inside of `impl` blocks for some insane reason.

ahh got it!

**Review by depombo — APPROVED** [review]

**Comment by depombo** [comment]
Pending tests passing
